main/main.c

```diff
@@ -1,35 +1,55 @@
-#include <stdio.h>
-#include "low_power_mode.h"
 #include "freertos/FreeRTOS.h"
 #include "freertos/task.h"
+#include "esp_err.h"
+#include "esp_log.h"
+
+#include "bme280.h"
+#include "low_power_mode.h"
 
-/**
- * @brief The task that enters into the low power mode.
- *
- * FreeRTOS task that enters into the low power mode. It sets the sleep time using the `low_power_mode_set_sleep_time` function and then enters into the low power mode using the `low_power_mode_enter_deep_sleep` function.
- */
-void low_power_mode_task()
+#define TAG_BME280 "BME280"
+#define TAG "MAIN"
+
+void main_task()
```

Comment:
Task w systemie FreeRTOS musi mieć zawsze pętlę nieskończoną while() lub for() - bez takiej pętli wynik jest nieprzewidywalny. Jedynym przykładem gdzie takiego czegoś nie robisz to jak na koniec task kasuje sam siebie (self).

Jednakże w każdym innym przypadku powinno to wyglądać tak:

```
/* Task to be created. */
void vTaskCode( void * pvParameters )
{
    /* The parameter value is expected to be 1 as 1 is passed in the
    pvParameters value in the call to xTaskCreate() below. 
    [configASSERT](https://www.freertos.org/a00110.html#configASSERT)( ( ( uint32_t ) pvParameters ) == 1 );

    for( ;; )
    {
        /* Task code goes here. */
    }
}
```

Najprawdopodobniej nie widzieliście problemów w działaniu, ponieważ po deep_sleep() system tak naprawdę się rebootuje więc cały FreeRTOS wstaje od nowa i tworzy taski też od nowa. Więc kod w main_task() nigdy nie wykonywał się do końca (nie wykonywał na końcu taska returna).

Reply:
```c  
ESP_LOGI(TAG_BME280, "BME280 init success");
for (int i = 1;; i++)
{
    ESP_LOGI(TAG_BME280, "Temperature: %.2f C, Pressure: %.2f hPa, Humidity: %.2f %%", bme280_read_temperature(), bme280_read_pressure() / 100, bme280_read_humidity());
    vTaskDelay(1000 / portTICK_PERIOD_MS);
    if (i % 10 == 0)
    {
        low_power_mode_set_sleep_time(10);
        low_power_mode_enter_deep_sleep();
    }
}
```

coś takiego może być? Nie wiem czy tak powinno się pisać bo na instynkt trochę śmierdzi pętla bez warunku

Reply:
Tak to jest ok. W takich małych systemach operacyjnych jak FreeRTOS koncepcja 'task' jest zrealizowana przez funkcję z nieskończoną pentlą. Kernel zajmuje się przełączaniem tych zadań.